File: backend/scanner-core/rule_pack_corpus/python/filesystem_vulnerable.py
"""Expected: 3 findings - write, read, destructive, one per tool."""

import os


@mcp.tool()
def write_file(path: str, content: str) -> str:
    """Writes content to a file."""
    with open(path, "w") as f:
        f.write(content)
    return "ok"


@mcp.tool()
def read_file(path: str) -> str:
    """Reads a file."""
    with open(path) as f:
        return f.read()


@mcp.tool()
def delete_file(path: str) -> str:
    """Deletes a file."""
    os.remove(path)
    return "deleted"

File: backend/scanner-core/rule_pack_corpus/typescript/filesystem_vulnerable.ts
// Expected: 3 findings - write, read, destructive, one per tool.

import * as fs from "fs";

server.registerTool("write_file", {
  description: "Writes content to a file",
  inputSchema: { path: z.string(), content: z.string() }
}, async ({ path, content }) => {
  fs.writeFileSync(path, content);
  return { content: [{ type: "text", text: "ok" }] };
});

server.registerTool("read_file", {
  description: "Reads a file",
  inputSchema: { path: z.string() }
}, async ({ path }) => {
  const data = fs.readFileSync(path);
  return { content: [{ type: "text", text: data.toString() }] };
});

server.registerTool("delete_file", {
  description: "Deletes a file",
  inputSchema: { path: z.string() }
}, async ({ path }) => {
  fs.unlinkSync(path);
  return { content: [{ type: "text", text: "deleted" }] };
});
